doc/man7/password-handling.pod

```diff
@@ -0,0 +1,89 @@
+=pod
+
+=head1 NAME
+
+password handling - How diverse parts of OpenSSL treat pass phrases
+
+=head1 DESCRIPTION
+
+=head2 The general case
+
+In a modern world with all sorts of character encodings, the treatment of pass
+phrases has become increasingly complex.
+
+The OpenSSL library doesn't treat pass phrases in any special way as a general
+rule, and trusts the application or user to choose a suitable character set
+and stick to that throughout the life of affected objects.
+This means that for an object that was encrypted using a pass phrase encoded in
+ISO-8859-1, that object needs to be decrypted using a pass phrase encoded in
+ISO-8859-1.
+Anything else will have the decryption fail.
+
+=head2 PKCS#12
+
+PKCS#12 is a bit different regarding pass phrase encoding.
+The standard stipulates that the pass phrase shall be encoded as a BMPString,
+i.e. UTF-16.
+
+OpenSSL tries to adapt to this requirements in one of the following manners:
+
+=over 4
+
+=item 1.
+
+Tries to read the pass phrase as UTF-8 and re-encode that to UTF-16, or failing
+that...
+
+=item 2.
+
+Assumes that the pass phrase is encoded in ASCII or ISO-8859-X and
+opportunistically prepends each byte with a zero byte and uses the result
+as a BMPString.
+
+=back
+
+Note that historical OpenSSL versions did variant 2 only, and that is the
+reason why OpenSSL still does this, to be able to read files produced with
+older versions.
+
+It should be noted that this approach isn't entirely fault free.
+A passphrase encoded in ISO-8859-X could very well have a sequence such as
+0xC3 0xAF (which is the two characters "LATIN CAPITAL LETTER A WITH TILDE"
+and "MACRON" in ISO-8859-1 encoding), but would be misunterpreted as the
+perfectly valid UTF-8 encoded U+00EF (LATIN SMALL LETTER I WITH DIARESIS).
+
+=head2 OSSL_STORE
+
+L<ossl_store(7)> acts as a general interface to access all kinds of
+objects, potentially protected with a pass phrase, a PIN or something else.
+This API currently doesn't stipulate any specific encoding of pass phrases,
+but uses the underlying routines with their behaviours.
+This means that the pass phrase to unlock a PKCS#12 file will be treated as
+described for PKCS#12 above, and the pass phrase for a PEM files will be
```

Comment:
This is descriptive but I think it would be useful to provide some prescriptive wording too — tell readers what they should be doing in order to use these APIs.

Include things like:
 • if you are opening a PKCS#12 file which may have been generated with either of the two OpenSSL behaviours described above *or* a correct implementation, and you have a password provided in a local charset which is neither UTF-8 and ISO8859-1, then you should try _<these>_ three ways to open it...
 • if you are opening a PEM file of unknown provenance and you are provided with a password in a local charset, you should try _<these>_ (two, I think) ways to open it. 

If you don't want the style of the document to be prescriptive HOWTO-like, that's fine. But please make sure it contains the information in an unambiguous form so that people can work out what they need to do for compatibility.

Reply:
Here's where examples help, again. For a given ISO8859-2 password you can show the actual BMP bytes used for the PKCS#12 key generation in all three cases, which helps to show what to use for attempted decryption of the file.

Reply:
Note that I have two more sections, CONCLUSIONS and newly added RECOMMENDATIONS, which I intend to use for exactly this.

Reply:
> If you are opening a PKCS#12 file which may have been generated with either of the two OpenSSL behaviours described above or a correct implementation, and you have a password provided in a local charset which is neither UTF-8 and ISO8859-1, then you should try three ways to open it...

Three?  Let's see, you can either provide a pass phrase in the original charset encoding, or in UTF-8 encoding...  what was the third option?

Reply:
You are in the ISO8859-2 locale. Your password is "ĂŻ" — that is, two characters:

    • U+0102 LATIN CAPITAL LETTER A WITH BREVE
    • U+017B LATIN CAPITAL LETTER Z WITH DOT ABOVE
Since you're in ISO8859-2, the password is represented as two bytes `0xC3 0xAF`

OpenSSL < 1.1 will use `00 C3 00 AF 00 00` to generate the key.
OpenSSL 1.1 will use `00 EF 00 00` to generate the key.

The third option is the *correct* option, `01 02 01 7B 00 00`.

Reply:
Or to turn that around to closer match your question _("you can provide a pass phrase in...")_... 

A well-behaved application would attempt to open a PKCS#12 file by providing the passphrase in three forms to OpenSSL 1.1.

 • Converted from the local charset to UTF-8. In the example above, this would be the bytes `c4 82 c5 bb` This will open *correct* files with `01 02 01 7b 00 00` for the key.
 • Mojibake "cast" from local charset to ISO8859-1 and then "converted" to UTF-8. In the example, take the ISO8859-2 bytes `C3 AF` and pretend they're ISO8859-1 and then convert *that* to UTF-8 to get `c3 83 c2 af`. That'll open files which used `00 c3 00 af 00 00` for the key.
 • If the native byte sequence in the local charset happens to be a valid non-ASCII UTF-8 sequence, then also mojibake-cast to UTF-8 and use that. I.e. pass the bytes to OpenSSL directly. That'll catch the (few, but non-zero) cases that OpenSSL 1.1 still doesn't get right, and in the example would use `00 EF 00 00` for generating the key.